### PR TITLE
GP updates

### PR DESCRIPTION
**Pull request by sbfnk**
This:

- introduces Gaussian Process updates at intervals of >1 day
- lints some of the GP stan code
- introduces Ornstein-Uhlenbeck and Matern 5/2 kernels
- tries to speed up computation by precomputing numerical constants

The wider spaced GP updates seem to work but results need more checking so I wouldn't suggest to merge this for now. The numerical precomputing didn't yield a drastic speed-up.

### CODE REVIEW

File: inst/stan/functions/gaussian_process.stan [hunk 1]
```diff
@@ -2,70 +2,112 @@
...
+  S = 2 * alpha * rho / (1 + rho^2 * w^2);
+  return S;
+}
+
 // spectral density of the Matern 3/2 kernel
-real spd_matern(real alpha, real rho, real w) {
+real spd_matern32(real alpha, real rho, real w) {
+  real S;
+  // S = 4 * alpha^2 * (sqrt(3) / rho)^3 * 1 / ((sqrt(3) / rho)^2 + w^2)^2;
+  S = 20.78461 * alpha / (rho^3 * (3 / rho^2 + w^2)^2);
```
Comment (seabbs) [inline inst/stan/functions/gaussian_process.stan:36]: I think stan is doing this for you already?

File: inst/stan/functions/rt.stan [hunk 2]
```diff
@@ -1,13 +1,33 @@
+// helper function to repeat each element of a vector `times` times,
+// up to a maximum of max_id`; if pad_zero is set to 1, a zero is
+// inserted at the beginning of vec
+vector rep_each(vector vec, int times, int max_id, int pad_zero) {
+  int vec_length = num_elements(vec);
+  vector[max_id] ret = rep_vector(0, max_id);
+  int remainder = max_id - (vec_length - 1) * times;
+  for (i in 1:(vec_length - 1)) {
+    for (j in 1:times) {
+      ret[j + (i - 1 + pad_zero) * times] = vec[i];
```
Comment (seabbs) [inline inst/stan/functions/rt.stan:10]: this is quite inefficient. I think a rewrite to use broadcasting might help?

File: inst/stan/functions/gaussian_process.stan (hunk 1 above)
Comment (sbfnk) [reply inst/stan/functions/gaussian_process.stan:36]: Well, that would explain the lack of performance improvement.

**Comment by sbfnk** [comment]
Closing as not currently intending to pursue this further.

**Comment by seabbs** [comment]
The GP code claims another victim. 2021 Sam 🖖🏻 you

**Pull request closed without merging**